Datadog/IntegrationUnitTests/Trace/HeadBasedSamplingTests.swift

```diff
@@ -0,0 +1,268 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import XCTest
+@testable import DatadogTrace
+import DatadogInternal
+import TestUtilities
+
+private class InstrumentedSessionDelegate: NSObject, URLSessionDataDelegate {}
+
+class HeadBasedSamplingTests: XCTestCase {
+    private var core: DatadogCoreProxy! // swiftlint:disable:this implicitly_unwrapped_optional
+    private var traceConfig: Trace.Configuration! // swiftlint:disable:this implicitly_unwrapped_optional
+
+    override func setUp() {
+        super.setUp()
+        core = DatadogCoreProxy()
+        traceConfig = Trace.Configuration()
+    }
+
+    override func tearDown() {
+        core.flushAndTearDown()
+        core = nil
+        traceConfig = nil
+        super.tearDown()
+    }
+
+    // MARK: - Local Tracing
+
+    // TODO: RUM-3470 Enable this test when head-based sampling is supported
+    func testSamplingLocalTrace() throws {
+        let localTraceSampling: Float = 50
+
+        // Given
+        traceConfig.sampleRate = localTraceSampling
+        Trace.enable(with: traceConfig, in: core)
+
+        // When
+        let parent = Tracer.shared(in: core).startSpan(operationName: "parent")
+        let child = Tracer.shared(in: core).startSpan(operationName: "child", childOf: parent.context)
+        let grandchild = Tracer.shared(in: core).startSpan(operationName: "grandchild", childOf: parent.context)
+        grandchild.finish()
+        child.finish()
+        parent.finish()
+
+        let spans = core.waitAndReturnSpanEvents()
+        XCTAssertEqual(spans.count, 3, "It must send all spans")
+
+        let allKept = spans.filter({ $0.isKept }).count == 3
+        let allDropped = spans.filter({ !$0.isKept }).count == 3
```

Comment:
❓ 
do we have inconsistent `isKept` flag for local spans in a trace?

Reply:
Yes 😞, this test is red if enabled. Although the [original issue](https://github.com/DataDog/dd-sdk-ios/issues/1713) touches only on head-based sampling for distributed traces I'll be fixing it for local traces as well. It might be one fix that works for both.